### PR TITLE
Add docs to register custom form extensions and types for unit testing

### PR DESCRIPTION
**Pull request by pierredup**
Docs for symfony/symfony#21780 to register custom extensions and types

### CODE REVIEW

File: form/unit_testing.rst [hunk 1]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
         }
     }
 
+Forms Using Validation
+------------------------------------
+
+If your forms uses the ``invalid_message`` or ``constraints`` option for validation, you need to
+register the validation extension which provides this options.
+Luckily Symfony provides a custom test class which does this for you.
+In order to have this option registered, your test needs to extend from the
```
Comment (atailouloute) [inline form/unit_testing.rst]: to extend the .. class would be better, no ? (Without from)

File: form/unit_testing.rst [hunk 2]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
         }
     }
 
+Forms Using Validation
+------------------------------------
```
Comment (OskarStark) [inline form/unit_testing.rst]: please keep the underline as long as the title

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: form/unit_testing.rst [hunk 3]
```diff
@@ -173,29 +192,22 @@ allows you to return a list of extensions to register::
...
-                new ValidatorExtension($this->validator),
+                new MyFormExtension(),
             );
         }
 
         // ... your tests
     }
 
+It is also possible to load custom form types, form type extensions or type guessers using the
+``getTypedExtensions``, ``getTypes`` and ``getTypeGuessers`` methods.
```
Comment (xabbuh) [inline form/unit_testing.rst]: This should be part of a separate pull request as we will have to merge it into the `master` branch (the rest can be merged into the docs for Symfony 2.7).

File: form/unit_testing.rst [hunk 4]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
         }
     }
 
+Forms Using Validation
+----------------------
+
+If your forms uses the ``invalid_message`` or ``constraints`` option for validation, you need to
```
Comment (HeahDude) [inline form/unit_testing.rst]: It should either be `If your form uses` or `If your forms use`.

File: form/unit_testing.rst [hunk 5]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
...
+register the validation extension which provides this options.
```
Comment (HeahDude) [inline form/unit_testing.rst]: `this` => `these`

File: form/unit_testing.rst [hunk 6]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
...
+Luckily Symfony provides a custom test class which does this for you.
```
Comment (HeahDude) [inline form/unit_testing.rst]: I would suggest `Luckily Symfony provides a base test class which takes care of it, just extend [class].`

And get rid of the following sentence, what do you think?

File: form/unit_testing.rst [hunk 7]
```diff
@@ -156,12 +156,31 @@ make sure the ``FormRegistry`` uses the created instance::
...
+In order to have these options registered, your test needs to extend the
+:class:`Symfony\\Component\\Form\\Tests\\Extension\\Validator\\Type\\TypeTestCase`
+class::
+
+    // tests/AppBundle/Form/Type/TestedTypeTests.php
```
Comment (HeahDude) [inline form/unit_testing.rst]: Please remove the last `s` in `TestedTypeTest.php`.

**Review by HeahDude — COMMENTED** [review]
Thanks for that PR! Can you please separate the changes as @xabbuh asked?

**Comment by pierredup** [comment]
I created a separate PR for 2.7 #7587.

When that is merged, it should squash this diff to only leave the sentence for the new methods added symfony/symfony#21780

**Comment by pierredup** [comment]
I removed the validator extension part in this docs, so it's not dependent on other changes, and can be merged to document the new methods added

File: form/unit_testing.rst (hunk 3 above)
Comment (xabbuh) [inline form/unit_testing.rst]: One thing I forgot: Can you also add a `versionadded` directive that mentions the introduction of the new methods?

File: form/unit_testing.rst [hunk 8]
```diff
@@ -199,6 +199,9 @@ allows you to return a list of extensions to register::
 It is also possible to load custom form types, form type extensions or type guessers using the
 ``getTypedExtensions``, ``getTypes`` and ``getTypeGuessers`` methods.
 
+.. versionadded:: 3.3
+    The ``getTypedExtensions``, ``getTypes`` and ``getTypeGuessers`` was introduced in Symfony 3.3.
```
Comment (HeahDude) [inline form/unit_testing.rst]: "The [...] methods were introduced [...]."

**Review by roman-1983 — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by HeahDude — APPROVED** [review]
👍

**Comment by xabbuh** [comment]
Thank you @pierredup. FYI, I have taken the proposed changes from symfony/symfony#22445 taken into account in 68bcaa0a7c11368325882236aa07c349258ab45e.

**Pull request closed without merging**